js/mbankScrapeRegular_onefile.js

```diff
@@ -0,0 +1,389 @@
+const ekonto = "";
+const lokaty = "";
+
+class AccountHistoryScraper {
+  constructor() {
+    this.n = 0;
+  }
+
+  transactions() {
+    return Array.from(
+      document.querySelectorAll("span.content-list-type-icon:not([data-original-title='Nierozliczone'])"));
+  }
+
+  openTransactionDetails(transactionEl) {
+    transactionEl.click();
+  }
+
+  getTransaction() {
+    const MEMO_COL = this.getElementByXpath("//tr[th[contains(text(), 'Rodzaj operacji')]]/td").innerHTML;
+
+    if (MEMO_COL == "PRZELEW REGULARNE OSZCZĘDZANIE") {
+      return new RegularneOszczedzanie();
+    }
+    else if (MEMO_COL == "ZAKUP PRZY UŻYCIU KARTY") {
+      return new ZakupKarta();
+    }
+    else if (MEMO_COL.includes("SPŁATA KARTY")) {
+      return new SplataKarty();
+    }
+    else if (MEMO_COL == "PRZELEW WŁASNY") {
+      return new PrzelewWlasny();
+    }
+    else if (MEMO_COL.includes("PROWIZJA")) {
+      return new Prowizja();
+    }
+    else if (MEMO_COL.includes("MTRANSFER")) {
+      return new Mtransfer;
+    }
+    else if (MEMO_COL.includes("BLIK - ZAKUP")) {
+      return new Blik();
+    }
+    else if (MEMO_COL.includes("PRZYCHODZĄCY")) {
+      return new PrzelewPrzychodzacy();
+    }
+    else if (MEMO_COL.includes("WYCHODZĄCY") && ! MEMO_COL.includes("MTRANSFER") ) {
+      return new PrzelewWychodzacy();
+    }
+    else if (MEMO_COL.includes("MOKAZJE")) {
+      return new Mokazje();
+    }
+    else if (MEMO_COL.includes("PODATEK")) {
+      return new Podatek();
+    }
+    else if (MEMO_COL.includes("KAPITALIZACJA")) {
+      return new Kapitalizacja();
+    }
+    else if (MEMO_COL.includes("WYPŁATA")) {
+      return new Wyplata();
+    }
+    else if (MEMO_COL.includes("WPŁATA")) {
+      return new Wplata();
+    }
+    else {
+      this.n+=1;
+      return console.log(this.n + ". Didn't recognize: " + MEMO_COL);
+    }
+  }
+
+  saveDetailsForCSV(arr) {
+    arr.push(this.getTransaction().toArray());
+  }
+
+  parseToCSVAndSaveFile(arr) {
+    let csvContent = "data:text/csv;charset=utf-8,";
+    arr.forEach(function(rowArray){
+      let row = rowArray.join(",");
+      csvContent += row + "\r\n";
+    });
+    let encodedUri = encodeURI(csvContent);
+    const link = document.createElement("a");
+    link.setAttribute("href", encodedUri);
+    link.setAttribute("download", "YNAB_ready" + Date.now() + ".csv");
+    link.click();
+  }
+
+  wait(ms) {
+    return new Promise((resolve) => setTimeout(resolve, ms));
+  }
+
+  getElementByXpath(path) {
+    return document.evaluate(path, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
+  }
+
+  async start() {
+    const arr = [["DATE","MEMO","PAYEE","AMOUNT"]];
+    // need to initialize transactions now
+    const transactionsArr = this.transactions();
+
+    for (let transactionEl of transactionsArr) {
+      // 1. open each transaction to enable selectors
+      this.openTransactionDetails(transactionEl);
+      await this.wait(5000);
```

Comment:
What about having something like `waitForElement` instead of hard-coded `wait (specific time)`? Here's how it could look like:

```js
waitForElement(selector) {
  return new Promise(resolve => {
    const interval = setInterval(() => {
      if (document.queryElement(selector)) {
        clearInterval(interval);
        resolve();
      }
    }, 300);
  });
}
```